### ISSUE TITLE
Issue connecting with just a metrics_api_key, Python KeyError: 'builds'

### ISSUE THREAD

**Issue by j0shd** [signalfx/collectd-jenkins#5]
Hi, I am having a couple of challenges with the collectd-jenkins integration for Signalfx.

I installed the metrics plugin in Jenkins and generated an API key.  Dropped python.py under /usr/share/collectd/collectd-jenkins

Setup collectd jenkins.conf following the outlined example:
```
LoadPlugin python
<Plugin python>
    ModulePath "/usr/share/collectd/collectd-jenkins"
    Import jenkins
    <Module jenkins>
        Host "localhost"
        Port "8080"
        MetricsKey "removed-key"
        Interval 60
    </Module>
</Plugin>
```
When I startup collectd I get errors stating invalid password/token for user:
```
[2018-01-26 15:55:06] signalfx-metadata: found host dom-use-devcimaster-05
[2018-01-26 15:55:06] Error making API call (HTTP Error 401: Invalid password/token for user:) http://localhost:8080/metrics/removed-key/ping/
[2018-01-26 15:55:06] Error making API call (HTTP Error 401: Invalid password/token for user:) http://localhost:8080/computer/api/json/
[2018-01-26 15:55:06] Unable to get data from http://localhost:8080/computer/api/json/ for computer
[2018-01-26 15:55:06] Error making API call (HTTP Error 401: Invalid password/token for user:) http://localhost:8080/metrics/removed-key/metrics/
[2018-01-26 15:55:06] Unable to get data from http://localhost:8080/metrics/removed-key/metrics/ for metrics
[2018-01-26 15:55:06] Error making API call (HTTP Error 401: Invalid password/token for user:) http://localhost:8080/metrics/removed-key/healthcheck/
[2018-01-26 15:55:06] Unable to get data from http://localhost:8080/metrics/removed-key/healthcheck/ for healthcheck
[2018-01-26 15:55:06] Error making API call (HTTP Error 401: Invalid password/token for user:) http://localhost:8080/api/json/
[2018-01-26 15:55:06] Unable to get data from http://localhost:8080/api/json/ for jenkins
[2018-01-26 15:55:06] signalfx-metadata: adding small dither of 41 seconds before sending notifications
```
However, I can get data from the provide URLs without needing a Jenkins user, including from the metrics plugin:
```
$ curl http://localhost:8080/computer/api/json/
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  3181  100  3181    0     0   997k      0 --:--:-- --:--:-- --:--:-- 1035k
{"_class":"hudson.model.ComputerSet","busyExecutors":0,"computer":[{"_class":"hudson.model.Hudson$MasterComputer","actions":[],"displayName":"master","executors":[{},{},{}],"icon":"computer.png","iconClassName":"icon-computer",...............

$ curl http://localhost:8080/metrics/removed-key/metrics/
{"version":"3.0.0","gauges":{"jenkins.executor.count.value":{"value":15},"je.........
```
Also, when I do configure jenkins.conf with a valid Jenkins User/API Token the access related issue is resolved but then a python exception comes up in the logs.
```
[2018-01-26 16:06:25] Unhandled python exception in read callback: KeyError: 'builds'
[2018-01-26 16:06:25] Traceback (most recent call last):
[2018-01-26 16:06:25]   File "/usr/share/collectd/collectd-jenkins/jenkins.py", line 524, in read_metrics
    read_and_post_job_metrics(module_config, module_config['base_url'], job['name'], last_timestamp)
[2018-01-26 16:06:25]   File "/usr/share/collectd/collectd-jenkins/jenkins.py", line 326, in read_and_post_job_metrics
    if resp_obj and resp_obj['builds']:
[2018-01-26 16:06:25] KeyError: 'builds'
[2018-01-26 16:06:25] read-function of plugin `python.localhost:8080' failed. Will suspend it for 120.000 seconds.

```
Versions:
collectd-5.8.0.sfx0-0.el6.x86_64
signalfx-collectd-plugin-0.2.26-sfx0.x86_64
collectd-python-5.8.0.sfx0-0.el6.x86_64
jenkins-2.73.2-1.1.noarch

Guessing the key error could be related to a version change in the Jenkins API.

Hopefully I am not missing something simple here, any suggestions are appreciated!

**Comment by nishantagarwal1990** [comment]
Hi j0shd,

Could you please provide some more details on the user's configuration? Is the security configured per project, etc.? 

Some more details would help a lot.

**Comment by j0shd** [comment]
I have the 'Matrix Authorization Strategy' plugin enabled under global security.  The user I was testing with has all privilleges for 'overall/credentials/agent/job/run/view/scm/metrics.'  project, and build, based security is not enabled.

On the jenkins server and can get all the build numbers for a project in JSON using curl, without providing user credentials.
```
$ curl http://localhost:8080/job/czen/job/CAREUS-33043/api/json?pretty=true&tree=builds[number,status,timestamp,id,result]
# this returns results like
$ {
  "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowJob",
  "actions" : [
    {

    },
    {

    },
    {

    },
    {
      "_class" : "com.cloudbees.plugins.credentials.ViewCredentialsAction"
    }
  ],
  "description" : null,
  "displayName" : "CAREUS-33043",
  "displayNameOrNull" : null,
  "fullDisplayName" : "czen Â» CAREUS-33043",
  "fullName" : "czen/CAREUS-33043",
  "name" : "CAREUS-33043",
  "url" : "https://jenkins.use.dom.carezen.net/job/czen/job/CAREUS-33043/",
  "buildable" : true,
  "builds" : [
    {
      "_class" : "org.jenkinsci.plugins.workflow.job.WorkflowRun",
      "number" : 17,
      "url" : "https://jenkins.use.dom.carezen.net/job/czen/job/CAREUS-33043/17/"
    },
```

Thanks for your reply!

**Comment by nishantagarwal1990** [comment]
Thank you for the reply. Could you provide a list of the plugins installed? Although i have the Matrix Authorization Strategy plugin installed, I do not see a section for credentials, project and build. 

Some snapshots of the configuration would also help. Please provide the versions of the plugins as well.

**Comment by j0shd** [comment]
This is a list of all the plugins installed and a screen shot of the matrix auth plugin.

[jenkins-plugins-installed.txt](https://github.com/signalfx/collectd-jenkins/files/1682047/jenkins-plugins-installed.txt)
![project-based-matrix-auth-plugin](https://user-images.githubusercontent.com/35851632/35629076-b38684c2-066b-11e8-8463-3dd1bb4e1f8f.JPG)

**Comment by j0shd** [comment]
Hi Nishant,

The fix in pull request #7, to allow anonymous reads, seems to work perfectly.   I am using jenkins.py from the develop branch.  I can now get both the JVM and job metrics with collectd-jenkins.

I also tried with providing a username/API key and the "KeyError: 'builds'" exception no longer comes up in the log.

Thanks for your help on this!

**Comment by nishantagarwal1990** [comment]
The bug has been fixed ion #7. Will close the issue here.
